typescript/web/src/components/core/combo-box/combo-box.context.tsx

```diff
@@ -0,0 +1,229 @@
+import { useControllableState } from "@chakra-ui/react";
+import {
+  useCombobox as useDownshift,
+  UseComboboxProps as DownshiftProps,
+  UseComboboxReturnValue as DownshiftReturnValue,
+  UseComboboxStateChange as DownshiftStateChange,
+} from "downshift";
+import { isEmpty } from "lodash/fp";
+import {
+  createContext,
+  PropsWithChildren,
+  useCallback,
+  useContext,
+  useMemo,
+  useState,
+} from "react";
+import { compareTwoStrings } from "string-similarity";
+
+export type DefaultComboBoxCompareKey = "text";
+
+export type ObjectKey = string | number | symbol;
+
+export type ComboBoxItem<TKey extends ObjectKey = DefaultComboBoxCompareKey> = {
+  [key in TKey | "id"]: string;
+};
+
+export type ComboBoxItemComponent<
+  TValue extends ComboBoxItem<TCompareKey>,
+  TCompareKey extends keyof TValue
+> = (value: TValue) => JSX.Element;
+
+export type CoreComboBoxProps<
+  TValue extends ComboBoxItem<TCompareKey>,
+  TCompareKey extends keyof TValue
+> = Omit<
+  DownshiftProps<TValue>,
+  "selectedItem" | "inputValue" | "onInputValueChange"
+>;
+
+export type ComboBoxProviderProps<
+  TValue extends ComboBoxItem<TCompareKey>,
+  TCompareKey extends keyof TValue
+> = CoreComboBoxProps<TValue, TCompareKey> & {
+  item: ComboBoxItemComponent<TValue, TCompareKey>;
+  listItem: ComboBoxItemComponent<TValue, TCompareKey>;
+  selectedItem?: TValue;
+  compareProp: keyof TValue;
+  onChange?: (value?: TValue) => void;
+};
+
+export type ComboBoxState<
+  TValue extends ComboBoxItem<ObjectKey> = Record<ObjectKey, string>,
+  TCompareKey extends keyof TValue = ObjectKey
+> = Omit<DownshiftReturnValue<TValue>, "selectedItem"> &
+  Pick<
+    ComboBoxProviderProps<TValue, TCompareKey>,
+    "items" | "onChange" | "item" | "listItem" | "selectedItem"
+  >;
+
+const ComboBoxContext = createContext({} as ComboBoxState);
+
+export const useComboBox = <
+  TValue extends ComboBoxItem<ObjectKey> = Record<ObjectKey, string>,
+  TCompareKey extends keyof TValue = ObjectKey
+>() =>
+  useContext(ComboBoxContext) as unknown as ComboBoxState<TValue, TCompareKey>;
+
+type FilterValuesOptions<
+  TValue extends ComboBoxItem<ObjectKey> = Record<ObjectKey, string>,
+  TCompareKey extends keyof TValue = ObjectKey
+> = Pick<ComboBoxState<TValue, TCompareKey>, "inputValue"> &
+  Pick<
+    ComboBoxProviderProps<TValue, TCompareKey>,
+    "items" | "compareProp"
+  > & {};
+
+const filterValues = <
+  TValue extends ComboBoxItem<ObjectKey> = Record<ObjectKey, string>,
+  TCompareKey extends keyof TValue = ObjectKey
+>({
+  items,
+  compareProp,
+  inputValue,
+}: FilterValuesOptions<TValue, TCompareKey>) => {
+  const lowerSearchText = inputValue.toLowerCase();
+  return items
+    .filter((item) => {
```

Comment:
They're complementary. If you don't use `compareTwoStrings` you can have cases where the search result would prioritise partial matches over exact matches for example.

Reply:
Ok. I'm just surprised to see that we need to use `.includes` as we had a heated debate about it weeks ago, but I understand now that `string-similarity` is only here to rank the results of the `includes`

Reply:
No, my problem was with the fact that it was supposedly too-complicated to include the string-similarity package in the loop 😉 

But still ; actually, it depends on the context. They are some edge-cases (e.g. _the code suggestions you have in VS Code_) where you'll prefer the filter to be more inclusive and allow typing errors and filter on a minimal score instead of using `.include`.

Reply:
I said that it looked overkill for the use we gonna make of it, and it still looks like it 😄

Reply:
Not that overkill actually if you compare with some other implementations 😉 
Here is the one used by the Qt framework (_the most used desktop GUI framework for cross-platform before Electron came-in_) for example: https://github.com/openwebos/qt/blob/master/src/gui/util/qcompleter.cpp

Reply:
But yes, the lesser, the better. We at-least agree on this 😅